### DISCUSSION TITLE
Little-Anti-Cheat Discussions

### DISCUSSION THREAD

**Discussion by J-Tanzanite** [J-Tanzanite/Little-Anti-Cheat#50] [General]
## Discussions :speaking_head: :speech_balloon: 
So Apparently this is a new feature on github, which is nice.
I think this is a perfect place for people to:
 * Ask questions
 * Talk about Lilac
 * Leave suggestions

And I can leave general updates about the project and the current situation.

If you have an issue, I still recommend you use the "issues" page.

**Comment by tappeh** [comment]
Is it possible to make the cheat detection logs go to like a mysql or somthing like that?

**Reply by J-Tanzanite** [reply]
It would be possible, the guy who originally suggested it asked if I could add it, but I don't know SQL so I couldn't do it...
But he offered to do it, but he later quit the project and since, that topic has been dead. :(

I'd like such a feature, as it would be VERY handy, but I don't know SQL and can't do it myself, sorry.

**Reply by azalty** [reply]
I don't know that much too, but I can try to do it. I usually create SQLite DB systems for my plugins since they have a faster response time (same server) and they're easier to setup (just add one in databases.cfg and you're done. no username/password)

However, I know the syntax for MySQL is a bit different than the one for SQLite and I'm unable to test MySQL DBs since I don't have one available rn and idk how to create one. I can only try to verify the syntax on some websites and check the doc, but I can certainly do SQLite.

A personal preference is to send each detection in a discord channel through a webhook. Since it's a webhook it doesn't require a bot / token and the rate limit seems pretty high (discord claims 10000 for 10 mins, which I don't believe is true, apparently it's like 30 messages/min, which is more than enough). Problem is it requires SteamWorks and Discord API by Deathknife (easy to install if downloads are provided directly, else it's a pain in the ass)

A way to ship these would be similar to this: https://github.com/rlevet/sm-no-dupe-account/releases

I'm okay to do that if you want, when I have some free time

**Reply by J-Tanzanite** [reply]
I wouldn't mind if you tried when you have time, but I wonder if it shouldn't be done inside Lilac directly.

The Discord request has been given to me recently, and it was originally in Lilac back when it was a private AC, but was removed before release.
Setting up Discord logging would be easy and neat.
Before I would add that tho, I should work on splitting up logging messages into different strings, so it doesn't send the entire thing (You wouldn't notice anything different in the logs, it's just an internal thing... Plus, it would make chat-messages of detections for admins in chat easier).
Plus, Discord has some nice formatting.

Anyway, when it comes to SQL logging.
It might be preferable if that was it's own plugin, and like the Discord plugin from Deathknife, exposed a native that Lilac could call.
Would be tidier for other plugins too I presume.

Sadly, like I mentioned earlier, I know nothing about SQL, and even if you did and managed to setup such a thing, I would have no way of actually testing it :/ 

So it all depends.
Would love if you could, but it's all... Complicated.

**Reply by azalty** [reply]
I don't know a thing about file splitting, but I don't see how you're going to pass all needed arguments efficiently through natives.

SQL queries will be asynchronous as well to prevent server freeze, so forwards are needed for callbacks, and again, it's dumb, because everything will be done on LilAC side, so no point on doing all that

Edit: since this is only about logging, either LilAC could call a native from the logging plugin and passing the message, steamid/client index or lilac could expose forwards that the logging plugin will get and log when they are called (idk if it's the case yet). No callbacks should be needed for that.

**Reply by J-Tanzanite** [reply]
My English isn't perfect, and I couldn't quite make sense of your message.

I assume your initial paragraph is referring to the string splitting and how Lilac would pass all those strings to the native.
It wouldn't work like that, although the strings are split, they would all get concatenated when doing file logging into a single long string. It would from there pass the entire string to the native.
Which is how it would work for the Discord plugin as well.
It seems like a weird thing to do, but it makes sense to do it like this when you consider in-game chat detections for admins.

The reason for this is pretty simple, logging the entire detection in-game chat for admins would be too long, so it has to be shortened; hence the splitting.

---

Although, there is another way to go about the in-game chat  detection messages for admins which I haven't thought about until now... Which... Would make the string splitting useless...

Anyway, I made the suggestion of SQL logging being it's own plugin in case other plugins wanna do logging through SQL as well. But if you don't think that's needed, then ok. ¯\\\_(ツ)_/¯
I don't know SQL so don't know what's best in terms of going about it.

**Reply by azalty** [reply]
Why would other plugins do SQL through LilAC's DB plugin?

Spoil: I'm not very good at English too

**Reply by J-Tanzanite** [reply]
"Lilac's DB"...
Wait, you're talking about SQL having a dedicated DB for lilac detections only?
...
Oh... Yeah now it makes more sense xD
I'm stupid :P

**Reply by azalty** [reply]
SQL Structure

-> Database
--> Table(s)
---> Data (rows)

We can use the default SQLite "sourcemod-local" database or create our own, we probably only need one table (a custom MySQL db is required in case you want to use MySQL, but again, if it's not going to be used by multiple servers at the same time it's not needed)

I don't see why any other plugin would want to call LilAC's natives

1. Plugins use their own SQL functions
2. No one needs to write/get data to/from LilAC's table, and if for some reason they need to, they can just write their own SQL functions
